Enable broker to destroy a VM and return the list of unfinished Cloudlets

pkoperek opened the pull request:
#### What's this Pull Request does, clear and precisely?

This PR adds a possibility to destroy a VM. When you request this, a list of affected cloudlets is returned so you can decide what to do with them.

#### Where should the reviewer start?

The example class is a good start: `SynchronousSimulationDestroyVmExample1`, most important changes are in `DatacenterBrokerAbstract` class. You can start there.

#### Are there some simulation examples to show the performed changes working?

See `SynchronousSimulationDestroyVmExample1`

#### What are the relevant issue ticket numbers?

https://github.com/manoelcampos/cloudsim-plus/issues/209

#### The issues or features included were discussed elsewhere outside github? If yes, please provide the links.

Nope

#### Any additional information you want to provide?

The sample simulation (`SynchronousSimulationDestroyVmExample1`) is based on the `SynchronousSimulationExample1` example.

The results are slightly different, however I think they are in line with expectations:

```
                                         SIMULATION RESULTS

Cloudlet|Status |DC|Host|Host PEs |VM|VM PEs   |CloudletLen|CloudletPEs|StartTime|FinishTime|ExecTime
      ID|       |ID|  ID|CPU cores|ID|CPU cores|         MI|  CPU cores|  Seconds|   Seconds| Seconds
-----------------------------------------------------------------------------------------------------
       0|SUCCESS| 1|   0|        4| 1|        2|      10000|          2|       20|        34|      14
       1|SUCCESS| 1|   0|        4| 1|        2|      10000|          2|        1|        23|      22
       2|SUCCESS| 1|   1|        4| 2|        2|      10000|          2|        2|        22|      20
       3|SUCCESS| 1|   1|        4| 3|        2|      10000|          2|        3|        23|      20
       4|SUCCESS| 1|   1|        4| 2|        2|      10000|          2|       24|        34|      10
       5|SUCCESS| 1|   0|        4| 1|        2|      10000|          2|        5|        27|      22
       6|SUCCESS| 1|   1|        4| 2|        2|      10000|          2|        6|        24|      18
       7|SUCCESS| 1|   1|        4| 3|        2|      10000|          2|        7|        25|      18
-----------------------------------------------------------------------------------------------------
```

One thing I'm not sure about is the fact that most of the cloudlets are processing for about 20s. I thought that with this setup a single cloudlet would fully occupy a VM and the next one would start after the first one had finished.

manoelcampos commented on cloudsim-plus/src/main/java/org/cloudbus/cloudsim/cloudlets/Cloudlet.java:
```diff
@@ -840,4 +840,6 @@
     @Override
     void setBroker(DatacenterBroker broker);
+
+    void reset();
```
```suggestion
    Cloudlet reset();
```

manoelcampos commented on cloudsim-plus/src/main/java/org/cloudbus/cloudsim/cloudlets/CloudletAbstract.java:
```diff
@@ -209,6 +198,23 @@ public CloudletAbstract(final long id, final long length, final long pesNumber)
         onUpdateProcessingListeners = new HashSet<>();
     }
 
+    public void reset() {
```
Since the method is called inside a constructor, set it to final to avoid "Constructor Calls Overridable Method" issue.

pkoperek replied:
:+1:

manoelcampos commented on cloudsim-plus/src/main/java/org/cloudbus/cloudsim/brokers/DatacenterBrokerAbstract.java:
```diff
@@ -771,14 +774,71 @@ public void requestIdleVmDestruction(final Vm vm) {
+    }
+
+    private List<Cloudlet> resetCloudletsFromVm(final Vm vm) {
+        final List<Cloudlet> cloudletsToReschedule = new ArrayList<>();
```
Probably the issue with your scenario with about ~60k cloudlets is here.
You are creating the ArrayList cloudletsToReschedule and adding elements to it.
When you call the first addAll, the list may be resized to accommodate all the new elements.
When you call addAll again, the list may be resized again.
Since we know exactly how many items will be added to the list, we can create it with the required capacity, avoiding the resize operation, as below:

```suggestion
        final int capacity = cloudletScheduler.getCloudletWaitingList().size() + cloudletScheduler.getCloudletExecList().size();
        final List<Cloudlet> cloudletsToReschedule = new ArrayList<>(capacity);
```

pkoperek commented on cloudsim-plus-examples/src/main/java/org/cloudsimplus/examples/SynchronousSimulationDestroyVmExample1.java:
```diff
@@ -0,0 +1,224 @@
+import java.util.function.Consumer;
+
+/**
+ * A example showing how to run the simulation synchronously,
```
:+1:

pkoperek commented on cloudsim-plus-examples/src/main/java/org/cloudsimplus/examples/SynchronousSimulationDestroyVmExample1.java:
```diff
@@ -0,0 +1,224 @@
+            if(iteration == 15) {
+                Vm vm = vmList.get(0);
+                List<Cloudlet> affected = broker0.requestVmDestruction(vm);
+                affected.stream().forEach(new Consumer<Cloudlet>() {
```
:+1:

pkoperek commented on cloudsim-plus/src/main/java/org/cloudbus/cloudsim/brokers/DatacenterBrokerAbstract.java:
```diff
@@ -771,14 +774,71 @@ public void requestIdleVmDestruction(final Vm vm) {
+        cloudlet.reset();
+    }
+
+    private Cloudlet extractResetCloudlet(CloudletExecution cloudletExecution) {
```
:+1:

pkoperek commented on cloudsim-plus/src/main/java/org/cloudbus/cloudsim/brokers/DatacenterBrokerAbstract.java:
```diff
@@ -771,14 +774,71 @@ public void requestIdleVmDestruction(final Vm vm) {
+    }
+
+    private List<Cloudlet> resetCloudlets(List<CloudletExecution> toReset) {
+        return toReset
```
:+1:

pkoperek commented on cloudsim-plus/src/main/java/org/cloudbus/cloudsim/brokers/DatacenterBrokerAbstract.java:
```diff
@@ -771,14 +774,71 @@ public void requestIdleVmDestruction(final Vm vm) {
+
+
+    public List<Cloudlet> requestVmDestruction(final Vm vm) {
+        if(getVmExecList().contains(vm)) {
```
Hmmm I actually wanted to check if this VM belongs to this datacenter... 
I guess I don't need that check actually - I'll use `isCreated`

pkoperek replied:
:+1:

pkoperek commented on cloudsim-plus/src/main/java/org/cloudbus/cloudsim/brokers/DatacenterBrokerAbstract.java:
```diff
@@ -771,14 +774,71 @@ public void requestIdleVmDestruction(final Vm vm) {
+            final List<Cloudlet> cloudletsFromVm = resetCloudletsFromVm(vm);
+            final List<Cloudlet> cloudlesSubmitted = new ArrayList<>();
+
+            for(Cloudlet cl : cloudletSubmittedList) {
```
:+1:

pkoperek commented on cloudsim-plus/src/main/java/org/cloudbus/cloudsim/brokers/DatacenterBrokerAbstract.java:
```diff
@@ -771,14 +774,71 @@ public void requestIdleVmDestruction(final Vm vm) {
+            }
+
+            cloudletSubmittedList.removeAll(cloudlesSubmitted);
+            cloudlesSubmitted.stream().forEach(this::resetCloudletStatus);
```
Actually a question... I just realized one thing: do I actually need the `resetCloudletsFromVm` method at all? If I understand correctly all cloudlets (even those which are executing) would be on the `cloudletSubmittedList` list? I guess I can remove `resetCloudletsFromVm` altogether?

pkoperek replied:
I did this refactoring - have a look and let me know what you think.

manoelcampos replied:
Yes, all cloudlets will be on `cloudletSubmittedList`. However, you can't use it because when you request the VM destruction, only running and waiting Cloudlets should be reset. Cloudlets already finished should not be touched.

pkoperek replied:
In the more recent version of code there is a test to check whether the cloudlets have been finished already: so any finished cloudlet should not be touched.

pkoperek commented on cloudsim-plus/src/main/java/org/cloudbus/cloudsim/brokers/DatacenterBrokerAbstract.java:
```diff
@@ -771,14 +774,71 @@ public void requestIdleVmDestruction(final Vm vm) {
+
+            vm.getHost().destroyVm(vm);
+            return Stream.concat(cloudletsFromVm.stream(), cloudlesSubmitted.stream()).collect(Collectors.toList());
+        } else {
```
:+1:

pkoperek commented on cloudsim-plus/src/main/java/org/cloudbus/cloudsim/brokers/DatacenterBrokerAbstract.java:
```diff
@@ -771,14 +774,71 @@ public void requestIdleVmDestruction(final Vm vm) {
+        return cloudletExecution.getCloudlet();
+    }
+
+    private void resetCloudletStatus(final Cloudlet cloudlet) {
```
:+1:

pkoperek commented:
hi @manoelcampos,

Fixed the issues you've found and just rebased the code. One thing I guess is worth mentioning is removing the `resetCloudletsFromVm`. I think given that all cloudlets are on the `cloudletSubmittedList` we probably don't even have to go through it. Does that sound ok? 
I've implemented this change as well and pushed to the MR.

manoelcampos commented on cloudsim-plus/src/main/java/org/cloudbus/cloudsim/brokers/DatacenterBrokerAbstract.java:
```diff
@@ -771,14 +774,37 @@ public void requestIdleVmDestruction(final Vm vm) {
         }
     }
 
+    public List<Cloudlet> requestVmDestruction(final Vm vm) {
```
Since the method requests the VM destruction right away, without even sending a message,
it should be renamed to `destroyVm`. This way, we provide name consistency with the `Host.destroyVm`.
It should be moved to DatacenterBroker interface too.

pkoperek replied:
:+1:  done

manoelcampos commented:
Hello @pkoperek 

I've merged your PR. Thanks again for your valuable contribution.
I just refactored `SynchronousSimulationDestroyVmExample1` and removed the line `affected.stream().forEach(cl -> cl.setSubmissionDelay(cl.getSubmissionDelay() + simulation.clock()))`. The submission delay is a relative value, so that you don't need to add the current clock (this is done inside the CloudSimEvent constructor).
Anyway, since the `Cloudlet.reset()` doesn't touch the `submissionDelay`, you don't need to reset it.

Since VM 0 is being destroyed at time 10, you can check that Cloudlets 0 and 4 (which were running on it) are resubmitted with the same delay: Cloudlet 0 with no delay (at tie 10) and Cloudlet 4 with 4 seconds of delay (at time 14).

> One thing I'm not sure about is the fact that most of the cloudlets are processing for about 20s. I thought that with this setup a single cloudlet would fully occupy a VM and the next one would start after the first one had finished.

Since we aren't explicitly defining a `CloudletScheduler` when creating the VMs, the default one is the `CloudletSchedulerTimeShared`. The behaviour you mentioned happens with the `CloudletSchedulerSpaceShared`.
